README.md
# Overwrite
Overwrite is a program that overwrites empty space on disk data and metadata.

**Documentation:** \
[https://ivoprogram.github.io?content=overwrite.html](https://ivoprogram.github.io?content=overwrite.html)

![Overwrite](images/overwrite-windows-ui.png)

The empty space on disks can contain sensitive data and information that can be recovered with recovery software and forensic, this can be security and privacy risk. \
To protect privacy and security the remaining data on empty space on disks has to be overwritten.

**Features**
- It is safe, does not overwrite existing files, and use operating system for overwrite.
- Can be used on different file systems and operating systems, as NTFS EXT4 FAT Windows Linux.
- Different options can be specified, as block size, number of files, quantity of data.
- It is free and open source with general public licence.

**Example**
- Write 10 metadata and 10Mb data, on Windows \
overwrite -meta:10 -data:10 -path:c:\

- Write random data, block size 512 bytes on Linux \
overwrite -rand -block:512 -data:10 -path:/mnt/usbdisk/

- Run directly overwrite.c source code with Tiny C compiler \
tcc -run overwrite.c -h

**Testing** \
Testing and issues for this project has been moved to link: \
https://github.com/ivoprogram/overwrite-testing

The collaborators are invited to contribute to this project. \
The objectives are to test the overwrite project, publish results, report issues. 

**Social Network** \
Facebook Page: https://facebook.com/OverwriteProgram \
Facebook Group: https://facebook.com/groups/2462527594073769 


![Overwrite](images/overwrite-windows-ui.png)
![Overwrite](images/overwrite-linux-ui.png)
![Overwrite](images/overwrite-linux.png)
![Overwrite](images/overwrite-unix-ufs.png)



changelog.txt


2019-10-04 1.0
Ivo Gjorgjievski
Overwrite program overwrite empty space on disk


2019-10-06 1.1
Ivo Gjorgjievski
Minor code changes, compiled also binaries for Unix (FreeBSD).


2019-10-28 1.2
Ivo Gjorgjievski
- The temporary files are written now in specified folder, not in subfolder.
- Removed temporary folder functions.
- Time calculation in hours, minutes, seconds, 00:00:00
- Add support for -files:all to overwrite all metadata.
- Minor code changes.
- Overwrite with user interface in C#.


2019-10-30 1.3
Ivo Gjorgjievski
- Now writing directories instead of small files, for efficency safety performance.
- Removed -files:all parameter, the metadata should not be filled all, side effects on NTFS.
- Renamed -files parameter with -dirs, now writing dirs for metadata overwrite.
- Fixed the progress bar refresh problem of user interface, overwrite-ui.
- Fixed -data:all parameter for overwrite-ui.


2019-11-02 1.4
Ivo Gjorgjievski
- Graphical user interface overwrite-ui.exe now work on Windows and Linux.
- When overwriting now first write data then metadata, to clean trace of data file.
- Fixed path buffer possible overflow from very long path, as reported by others.


2019-11-03 1.4.1
Ivo Gjorgjievski
- Changed path delimiter from \\ to / in overwrite.c for code to be similar with C# in Form1.cs, this is not bug, worked in windows and linux.
- Check path length moved from getpath() to parseargs(), this is not bug.
- Renamed Form1.cs to overwrite.cs, this is C# UI version of overwrite program.


2019-11-07 1.4.2
Ivo Gjorgjievski
- In overwrite.c fixed issues for special chars in path in Unix as reported in issues #2
- FILE_PREFIX changed in FILE_SUFFIX and set it to empty, used for testing only for long file names in FAT.
- In overwrite.cs fixed special chars in path as previous issue.
- In overwrite.cs added path variable, textPath.Text is not used and changed directly.


2019-11-09 1.5 
Ivo Gjorgjievski
- Changed parameter -dirs to -meta, and set mb as default, ex -data:10 = -data:10mb
- Set FILE_SUFFIX to 32 chars for overwrite long file names on some file sistems.
- Add flushing data buffers, in previous versions not worked correctly.


2019-11-14 1.5.1 
Ivo Gjorgjievski
- Fixed issue with metadata remanence on FAT, EXT.
- Metadata and data is written in destination dir not in subdir.
- Using small files instead of dirs for metadata and data cleaning.
- This changes were required to support different file systems and operating systems.


2019-11-16 1.5.2
Ivo Gjorgjievski
- Fixed issue with metadata remanence on FAT, EXT.
- Also upper case, and long file names are cleaned.


2019-11-20 1.5.3
Ivo Gjorgjievski
- Temporary files names are written in different pattern, fixed issue with metadata remanence on FAT, EXT on new tests.


2019-11-21 1.6
Ivo Gjorgjievski
- Temporary file names pattern set to normal, tests passed.
- User interface overwrite.cs updated same as overwrite.c
- Created release of this version


2019-11-21 1.6.1
Ivo Gjorgjievski
- Write different pattern in metadata file names.


c/overwrite.c

/*
TITLE:			Overwrite Program
DESCRIPTION:	Overwrite empty space on disk data and metadata.
CODE:			github.com/ivoprogram/overwrite
LICENSE:		GNU General Public License v3.0 http://www.gnu.org/licenses/gpl.html
AUTHOR:			Ivo Gjorgjievski
WEBSITE:		ivoprogram.github.io
VERSION:		1.6.1 2019-11-21

*/


// Include -----
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <time.h>
#include <limits.h>
#include <ctype.h>

#ifdef _WIN32	// _WINDOWS
#include <windows.h>	// HANDLE
#include <io.h>			// _get_osfhandle
#else			// Linux, Unix
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#endif


// Definitions -----
#define TRUE 1       
#define FALSE 0
#define SAFE_WRITE TRUE		// Check if file exists before overwrite
#define BLOCK_SIZE 4096		// Default block size, used in NTFS EXT4
#define PATH_LENGTH 4096	// Path buffer size
#define FILE_SUFFIX "xxxxx"		// File suffix
#define FILE_SUFFIX2 "X"	// File suffix
#define VERSION "Overwrite Version 1.6.1 2019-11-21"	// Program version


// Arguments structure -----
struct args {
	int res;				// Result from called function
	int test;				// For test, don't delete written files
	int one;				// Write one's
	int ran;				// Write random data
	int blocksize;			// Block size
	int files;				// Number of files
	int files2;				// Number of files written, for cleaning
	int data;				// Quantity of data to write
	char *dataunit;			// Data unit mb, gb, all, default mb
	char *path;				// Path where to write 
	char *pathend;			// Slash at the end of path, if missing
	char pathbuf[PATH_LENGTH + 1024];		// Buffer path + file + suffix 
	unsigned char *blockbuf;	// Block buffer init with calloc
};


// Headers -----
struct args parseargs(int argc, char *argv[]);
struct args writefiles(struct args argsv);
struct args writefiles2(struct args argsv, const char *suffix, int data);
struct args writedata(struct args argsv);
int writefile(char *filepath, unsigned char *buf, int bufsize);
void cleanfiles(struct args argsv, const char *suffix);
void cleandata(struct args argsv);
void freemem(struct args argsv);

void safecheck(struct args argsv);
void randbuf(unsigned char *buf, int bufsize);
void getpath(struct args *argsv, const char *name);
void syncfile(FILE *fp);
void printtime(clock_t ticks);
int strstart(const char *str, const char *sub);
int strend(const char *str, const char *sub);
int stricmp2(const char *str1, const char *str2, size_t count);

void errmsg(struct args argsv, char *message);
void exitmsg(struct args argsv, char *message);
void exiterr(struct args argsv, char *message);
void errcheckint(int res, struct args argsv, char *message);
void errcheckptr(void *res, struct args argsv, char *message);
void usage();


// Main function -----
int main(int argc, char *argv[])
{
	// Local vars
	clock_t ticks;
	struct args argsv;

	// Init 
	printf("\n");
	ticks = clock();

	// Parse arguments
	argsv = parseargs(argc, argv);

	// Write files
	argsv = writefiles(argsv);

	// Write data
	argsv = writedata(argsv);

	if(!argsv.test){ cleandata(argsv); }

	// Clean memory
	freemem(argsv);

	// Done
	ticks = clock() - ticks;
	printtime(ticks);
	printf("Done. \n");
	//system("pause");
	return 0;

}// 



// Parse arguments
struct args parseargs(int argc, char *argv[])
{
	// 
	int index = 0;			// Index
	int res = 0;			// Result
	struct args argsv;		// Arguments structure

	// Init arguments structure
	argsv.res = 0;
	argsv.test = 0;
	argsv.one = 0;
	argsv.ran = 0;
	argsv.blocksize = BLOCK_SIZE;
	argsv.files = 0;
	argsv.files2 = 0;
	argsv.data = 0;
	argsv.dataunit = "";
	argsv.path = "";
	argsv.pathend = "";
	argsv.blockbuf = NULL;
	strcpy(argsv.pathbuf, "");

	// Usage
	if(argc < 2){ usage(); exit(0); }

	// Check for space in args
	for(index = 0; index < argc; index++)
	{
		if(strend(argv[index], "-block:") ||
			strend(argv[index], "-meta:") ||
			strend(argv[index], "-data:") ||
			strend(argv[index], "-path:")
			){
				exitmsg(argsv, "Error, there are empty arguments.");
		}
	}

	// Parse arguments
	for(index = 0; index < argc; index++)
	{
		// 
		if(strstart(argv[index], "-h") || strstart(argv[index], "--help") ){ 
			usage();
			exit(0);
		}
		else if(strstart(argv[index], "-v") || strstart(argv[index], "--version") ){ 
			printf(VERSION "\n");
			exit(0);
		}
		else if(strstart(argv[index], "-test")){ 
			argsv.test = 1; 
		}
		else if(strstart(argv[index], "-one")){ 
			argsv.one = 1; 
		}
		else if(strstart(argv[index], "-rand")){ 
			argsv.ran = 1; 
		}
		else if(strstart(argv[index], "-block:")){
			argsv.blocksize = atoi(argv[index] + 7);
			if(argsv.blocksize == 0){ argsv.blocksize = BLOCK_SIZE; }
		}
		else if(strstart(argv[index], "-meta:")){
			argsv.files = atoi(argv[index] + 6);
		}
		else if(strstart(argv[index], "-data:")){
			argsv.data = atoi(argv[index] + 6);
			argsv.dataunit = argv[index] + 6;
		}
		else if(strstart(argv[index], "-path:")){
			argsv.path = argv[index] + 6;
		}

	}// for

	// Meta or data args are required
	if(argsv.files == 0 && strcmp(argsv.dataunit, "") == 0)
	{
		exitmsg(argsv, "Error, arguments -meta: or -data: are missing.");
	}

	// Path arg is required
	if(strcmp(argsv.path, "") == 0)
	{
		exitmsg(argsv, "Error, argument -path: is missing.");
	}

	// Check if path is less than PATH_LENGTH
	if(strlen(argsv.path) > PATH_LENGTH)
	{
		exitmsg(argsv, "Path is too long, parameter -path: \n");
	}

	// Add / at the end of path if missing, windows unix
#ifdef _WIN32	//  __linux__
	if(!strend(argsv.path, "/") && !strend(argsv.path, "\\")){
		argsv.pathend = "/";
	}
#else      
	if(!strend(argsv.path, "/")){
		argsv.pathend = "/";
	}
#endif

	// Initialize blockbuf with calloc, set with 0.
	argsv.blockbuf = (unsigned char*) calloc(argsv.blocksize, sizeof (unsigned char));
	errcheckptr(argsv.blockbuf, argsv, "Block buffer initialization error. ");

	// If one, set buffer with one's
	if(argsv.one){ memset(argsv.blockbuf, 255, argsv.blocksize);  }

	// If random set buffer with random data
	if(argsv.ran){ randbuf(argsv.blockbuf, argsv.blocksize); }

	argsv.res = 0;
	return argsv;

}//

// writefiles
struct args writefiles(struct args argsv)
{
	// If files 0 return
	if(!argsv.files){ argsv.res = 0; return argsv; }
	argsv.files2 = 0;	// Number of files written, for clean

	// 
	printf("Writing metadata: \n");

	// Write block size files for overwrite metadata and small files
	argsv = writefiles2(argsv, FILE_SUFFIX, TRUE);
	if(!argsv.test){ 
		cleanfiles(argsv, FILE_SUFFIX);
	}

	// Write empty files for overwrite upper case metadata entries
	argsv = writefiles2(argsv, FILE_SUFFIX2, FALSE);
	if(!argsv.test){ 
		cleanfiles(argsv, FILE_SUFFIX2); 
	}

	// 
	return argsv;

}// 


// writefiles
struct args writefiles2(struct args argsv, const char *suffix, int data)
{
	// 
	int index = 0;		// Index
	int res = 0;		// Response
	int cont = 0;		// Counter for file name
	float prog = 0.1F;	// Progress
	char stri[64];		// Dir name

	// Write files
	//printf("Writing metadata: \n");
	for(index = 1; index <= argsv.files; index++)
	{
		// Get path
		sprintf(stri, "%d%s", index, suffix);
		getpath(&argsv, stri);					

		// Safe write check
		safecheck(argsv);	

		// Create file block size, or empty
		if(data){
			res = writefile(argsv.pathbuf, argsv.blockbuf, argsv.blocksize);
		}
		else{
			res = writefile(argsv.pathbuf, argsv.blockbuf, 0);
		}

		errcheckint(res, argsv, "Error writing metadata file.");
		argsv.files2 = index;	// Number of files written, for clean

		// Statistics
		if(data && index > argsv.files * prog)
		{
			printf("%.0f%% ", prog * 100 );
			prog += 0.1F;

		}// if

	}// for

	if(data){ printf("100%% \n"); }
	argsv.res = 0;
	return argsv;

}//


// writefile
int writefile(char *filepath, unsigned char *buf, int bufsize)
{
	int resi;		// Response int
	size_t rest;	// Response t
	FILE *fp;		// File pointer

	// Open file 
	fp = fopen(filepath, "wb");
	if(fp == NULL){ return -1; }

	// Write one block
	if(bufsize > 0){
		rest = fwrite(buf, bufsize, 1, fp);
		if(rest != 1){ fclose(fp); return -1; }
	}

	// 
	syncfile(fp);
	resi = fclose(fp);
	if(resi == EOF){ return -1; }

	return 0;

}// 


// writedata
struct args writedata(struct args argsv)
{
	int blocks = 0;			// Number of blocks to write
	int blocksmb = 0;		// Blocks per megabyte
	int blocksgb = 0;		// Blocks per gigabyte
	int blockscont = 0;		// Blocks count for statistics
	int blockscont2 = 0;	// Blocks count 2 for statistics

	int index = 0;			// Index
	int resi = 0;			// Response int
	size_t rest = 0;		// Response size_t
	double prog = 0.1;		// Progress
	FILE *fp;				// File pointer

	// if data 0 return
	if(argsv.data < 1 && !strstart(argsv.dataunit, "all"))
	{ argsv.res = 0; return argsv; }


	// How many blocks to write, 1Mb / blocksize
	blocksmb = (1024 * 1000) / argsv.blocksize;
	blocksgb = (1024 * 1000 * 1000) / argsv.blocksize;
	blockscont = blocksgb;

	// Calculate blocks to write
	blocks = argsv.data * blocksmb; // Default mb
	if(strend(argsv.dataunit, "gb")) { blocks = argsv.data * blocksgb; }
	else if(strstart(argsv.dataunit, "all")) { blocks = INT_MAX; argsv.data = 1; }
	// Set data to 1 for clean

	// Get path
	getpath(&argsv, "0");
	strcat(argsv.pathbuf, FILE_SUFFIX);

	// Safe write check
	safecheck(argsv);

	// Open file
	fp = fopen(argsv.pathbuf, "wb");
	errcheckptr(fp, argsv, "Error writing data file.");

	// Write data
	printf("Writing data: \n");
	for(index = 0; index < blocks; index++)
	{
		// write block
		rest = fwrite(argsv.blockbuf, argsv.blocksize, 1, fp);

		// If error or disk full exit function
		if(rest != 1 && index > 0){ 
			printf("100%% \n"); 
		}

		if(rest != 1){ 
			syncfile(fp);
			fclose(fp);
			argsv.res = -1; 
			return argsv;
		}

		// Statistics
		if(blocks != INT_MAX && index > blocks * prog){
			printf("%.0f%% ", prog * 100 );
			prog += 0.1;

		}// if

		// Statistics for -data:all argument, up to 10gb
		if(blocks == INT_MAX && index >= blockscont && blockscont2 < 10){
			blockscont += blocksgb;
			blockscont2++;
			printf("%dGb ", blockscont2 );
		}

	}// for

	// Done
	printf("100%% \n");
	syncfile(fp);
	fclose(fp);
	argsv.res = 0;
	return argsv;

}// 


// Clean files
void cleanfiles(struct args argsv, const char *suffix)
{
	// 
	int index;		// Index
	int cont = 0;		// Counter for file name
	int res;		// Response
	char stri[64];	// Name
	FILE *fp;		// File pointer


	// For number of files written
	for(index = 1; index <= argsv.files2; index++){

		// Get path
		sprintf(stri, "%d%s", index, suffix);
		getpath(&argsv, stri);					

		// Remove file
		fp = fopen(argsv.pathbuf, "w");
		syncfile(fp);
		fclose(fp);
		res = remove(argsv.pathbuf);

	}// for

}//


// Clean data
void cleandata(struct args argsv)
{
	// 
	int res;		// Response
	char stri[64];	// Name
	FILE *fp;		// File pointer

	// If no data written return
	if(argsv.data == 0){ argsv.res = 0; }

	// File path
	sprintf(stri, "%d", 0);
	getpath(&argsv, stri);
	strcat(argsv.pathbuf, FILE_SUFFIX);

	// Remove data file
	fp = fopen(argsv.pathbuf, "w");
	syncfile(fp);
	fclose(fp);
	res = remove(argsv.pathbuf);

}//


// Free memory 
void freemem(struct args argsv)
{
	errno = 0;

	if(argsv.blockbuf != NULL){
		free(argsv.blockbuf);
		argsv.blockbuf = NULL;
	}

	if(errno != 0){
		printf("Error on free() memory: %d %s \n", errno, strerror(errno));
	}

}// 



// Util functions ---------------------

// Check if file exists before overwrite
void safecheck(struct args argsv)
{
	FILE *fp;	// File pointer

	// Check if file exists, secure overwrite
	if(SAFE_WRITE && !argsv.test && (fp = fopen(argsv.pathbuf, "r")) != NULL)
	{
		fclose(fp);
		printf("File exists, clean file before overwrite, safe write is on. \n%s \n", 
			argsv.pathbuf);
		exit(0);
	}

}// 


// Set buffer with random data
void randbuf(unsigned char *buf, int bufsize)
{
	int index;
	srand((unsigned)time(NULL));	// Init rand

	for(index = 0; index < bufsize - 1; index++){
		buf[index] = rand() % 256; // 
	}

}// 


// Get path, slash, file name
void getpath(struct args *argsv, const char *name)
{
	// Add to buffer path, slash, prefix
	strcpy(argsv->pathbuf, argsv->path);
	strcat(argsv->pathbuf, argsv->pathend);
	strcat(argsv->pathbuf, name);

}// 


// Sync file buffer to disk, windows linux
void syncfile(FILE *fp)
{

#ifdef _WIN32

	int res = 0;
	int fd = 0; 
	HANDLE hd = NULL;

	fd = _fileno(fp);
	hd = (HANDLE) _get_osfhandle(fd);
	res = fflush(fp);
	res = FlushFileBuffers(hd);
	//printf("%d \n", res);

#else	//  __linux__

	int fd = fileno(fp);
	fflush(fp);
	fsync(fd);

#endif

}//


// Print time
void printtime(clock_t ticks)
{
	// Total seconds
	float total = (float) ticks / CLOCKS_PER_SEC;

	// Hours, minutes, seconds
	int hours = (int)total / 60 / 60;
	int minutes = (int)total / 60 % 60;
	int seconds = (int)total % 60;

	// Milliseconds
	int intpart = (int)total;
	float decpart = total - intpart;
	float milsec = decpart * 1000;

	printf("\nTime: %02d:%02d:%02d.%.0f \n", 
		hours, minutes, seconds, milsec);

}// 


// String start with, 0 false, 1 true
int strstart(const char *str, const char *sub)
{
	return stricmp2(str, sub, strlen(sub)) == 0;
}


// String end with, 0 false, 1 true
int strend(const char *str, const char *sub)
{
	int sublen = (int)strlen(sub);
	return stricmp2(str + strlen(str) - sublen, sub, sublen) == 0;
}


// String compare case insensitive, windows linux
int stricmp2(const char *str1, const char *str2, size_t count)
{
#ifdef _WIN32	
	return _strnicmp(str1, str2, count);
#else	//  __linux__
	return strncasecmp(str1, str2, count);
#endif
}


// Print error message, do not exit
void errmsg(struct args argsv, char *message)
{
	printf("Error %d, %s. \n%s \n", errno, strerror(errno), message);
}// 


// Exit with message
void exitmsg(struct args argsv, char *message)
{
	printf("%s \n", message);
	freemem(argsv);
	exit(0);

}// 


// Exit with error message
void exiterr(struct args argsv, char *message)
{
	printf("Error %d, %s. \n%s \n", errno, strerror(errno), message);
	freemem(argsv);
	exit(-1);

}// 


// Check error integer
void errcheckint(int res, struct args argsv, char *message)
{
	if(res == -1){ exiterr(argsv, message); }
}// 


// Check error pointer
void errcheckptr(void *res, struct args argsv, char *message)
{
	if(res == NULL){ exiterr(argsv, message); }
}// 


// 
void usage()
{
	printf(
		"Overwrite empty space on disk data and metadata.\n"
		VERSION "\n\n"
		"USAGE: \n"
		"overwrite [-h -v -test -one -rand -block:] (-meta: &| -data:) -path: \n"
		"  -h \t\t Print help and usage message. \n"
		"  -v \t\t Print program version. \n"
		"  -test \t For test, don't delete written files. \n"
		"  -one \t\t Overwrite with 1, default with 0. \n"
		"  -rand \t Overwrite with random data, default with 0. \n"
		"  -block \t Block size, default 4096 bytes as NTFS and EXT4. \n"
		"  -meta \t Number of metadata to write, ex: 1, 10.\n"
		"  -data \t Quantity of data to write, 1mb, 1gb, all, default mb. \n"
		"  -path \t Path to directory where to write. \n\n"
		"EXAMPLE: \n"
		"- Write 10 metadata and 10Mb data, on Windows \n"
		"overwrite -meta:10 -data:10 -path:c:\\ \n"
		"- Write random, block 512 bytes, 10Mb data, on Linux \n"
		"overwrite -rand -block:512 -data:10 -path:/mnt/usbdisk/ \n\n"
		"Copyright GPLv3 http://github.com/ivoprogram/overwrite \n"

		);

}// 




